### PR TITLE
Prevent changing DS fields that would invalidate its SSL key

### LINKED ISSUE #2805: Traffic Ops Does Not Update DNSSEC Keys on Delivery Service Update
Traffic Ops Does Not Update DNSSEC Keys on Delivery Service Update.

As far as I can tell, this has been a bug forever. The old Perl endpoint doesn't update them either, and it appears it never has.

This means updating anything in a DS which changes the served FQDNs (e.g. Routing Name, Regexes) will break DNSSEC.

### LINKED ISSUE #5075: Invalid ssl keys when the cdn of a delivery service is changed
## I'm submitting a ...
-  bug report

## Traffic Control components affected ...

-  Traffic Ops
-  Traffic Ops ORT
-  Traffic Portal
-  Traffic Router
-  Traffic Vault

## Current behavior:
a ds w/ ssl keys has ssl keys of 1 of 3 types:
- self-signed ssl keys generated via the TO API generate endpoint
- ssl keys generated via the Lets Encrypt TO API generate endpoint
- ssl keys that were manually uploaded (that were signed/generated via a 3rd party)

when switching the cdn of a ds, 2 things must happen:
1. the ssl keys need to be moved from cdn1 to cdn2 (this is done with #4988 )
2. the ssl keys need to be potentially regenerated / uploaded to account for the modified DS fqdn due to the cdn change.

## Expected behavior:
Up for debate.

## Minimal reproduction of the problem with instructions:
- Create a DS in cdn1 and set protocol to `http and https`. Generate self-signed ssl keys. Queue updates and snapshot. Test the DS.
- Change the DS cdn to cdn2. Snap and queue updates. Test the DS.

### PR DESCRIPTION
**Pull request by shamrickus**
## What does this PR (Pull Request) do?

- [x] This PR fixes #5075, #1847 and partially fixes #2805 by disallowing changes to the DS that would affect the SSL key.

## Which Traffic Control components are affected by this PR?

- Traffic Ops
- Traffic Portal
- Traffic Vault

## What is the best way to verify this PR?
Try to change the routing name/CDN on a DS with an ssl key in both TO and TP to verify that it is not permitted.

## If this is a bug fix, what versions of Traffic Control are affected?

- master (435a8b0)

## The following criteria are ALL met by this PR

- [x] This PR includes tests OR I have explained why tests are unnecessary
- [x] This PR includes documentation OR I have explained why documentation is unnecessary
- [x] This PR includes an update to CHANGELOG.md OR such an update is not necessary
- [x] This PR includes any and all required license headers
- [x] This PR ensures that database migration sequence is correct OR this PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

### CODE REVIEW

File: traffic_portal/app/src/common/modules/form/deliveryService/edit/FormEditDeliveryServiceController.js [hunk 1]
```diff
@@ -161,6 +166,30 @@ var FormEditDeliveryServiceController = function(deliveryService, origin, topolo
 	};
 
 	$scope.save = function(deliveryService) {
+		if (deliveryService.sslKeyVersion !== null && deliveryService.sslKeyVersion !== 0 &&
```
Comment (mattjackson220) [inline traffic_portal/app/src/common/modules/form/deliveryService/edit/FormEditDeliveryServiceController.js:169]: I dont think this page will ever have deliveryService.sslKeyVersion !== null. I think that /deliveryservices endpoint doesnt return the ssl key version.  I dont think we know that until we are onto the ssl keys page itself.  TP still fails to updated the DS but just returns a Forbidden message

**Comment by mattjackson220** [comment]
Should we add a note or something in the docs to let people know that you cannot update these fields if SSL keys are present?

File: traffic_portal/app/src/common/modules/form/deliveryService/edit/FormEditDeliveryServiceController.js (hunk 1 above)
Comment (shamrickus) [reply traffic_portal/app/src/common/modules/form/deliveryService/edit/FormEditDeliveryServiceController.js:169]: From my testing, the ssl_key_version on the deliveryservice table is always updated when the ssl keys are changed/created. Also the /deliveryserivces endpoint does return that field as `sslKeyVersion`. Inspecting the network call in TP when loading the DS table, each DS has it's `sslKeyVersion` returned.

Comment (mattjackson220) [reply traffic_portal/app/src/common/modules/form/deliveryService/edit/FormEditDeliveryServiceController.js:169]: interesting. maybe cause i was just running TO and TP locally. ill try again in a better environment

Comment (mattjackson220) [reply traffic_portal/app/src/common/modules/form/deliveryService/edit/FormEditDeliveryServiceController.js:169]: my bad. definitely just my setup. looks good in ciab.

File: traffic_portal/conf/config.js [hunk 2]
```diff
@@ -21,27 +21,27 @@
 module.exports = {
     timeout: '120s',
     useSSL: true, // set to true if you plan to use https (self-signed or trusted certs).
-    port: 80, // set to http port
-    sslPort: 443, // set to https port
+    port: 880, // set to http port
```
Comment (mattjackson220) [inline traffic_portal/conf/config.js:24]: are these changes leftover from local testing?

Comment (shamrickus) [reply traffic_portal/conf/config.js:24]: -sigh-, yes

**Comment by zrhoffman** [comment]
The DeliveryServices tests fail for me in `SSLDeliveryServiceCDNUpdateTest()` `createBlankCDN()` when `TestDeliveryServices` is run by itself (after `TestMain` is run to initialize the API tests) and `includeSystemTests` is enabled:

```go
go test ./v3 -test.v -cfg=../conf/traffic-ops-test.conf -fixtures=tc-fixtures.json -test.run 'TestMain$|^TestDeliveryServices$'
```

<details><summary>Log</summary>

```go
INFO: traffic_ops_test.go:71: 2020-10-06T13:31:19.364733051-06:00: Using Config values:
			   TO Config File:       ../conf/traffic-ops-test.conf
			   TO Fixtures:          tc-fixtures.json
			   TO URL:               https://localhost:6443
			   TO Session Timeout In Secs:  60
			   DB Server:            localhost
			   DB User:              traffic_ops
			   DB Name:              traffic_ops
			   DB Ssl:               false
               UseIMS:               true
=== RUN   TestDeliveryServices
    cdns_test.go:87: Response:  {[{cdn was created. success}]}
    cdns_test.go:87: Response:  {[{cdn was created. success}]}
    cdns_test.go:87: Response:  {[{cdn was created. success}]}
    cdns_test.go:87: Response:  {[{cdn was created. success}]}
    types_test.go:87: ---- CreateTestTypes ----
    user_test.go:101: Response:  {[{user was created. success}]}
    user_test.go:101: Response:  {[{user was created. success}]}
    user_test.go:101: Response:  {[{user was created. success}]}
    user_test.go:101: Response:  {[{user was created. success}]}
    user_test.go:101: Response:  {[{user was created. success}]}
    user_test.go:101: Response:  {[{user was created. success}]}
    user_test.go:101: Response:  {[{user was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    parameters_test.go:75: Response:  {[{param was created. success}]}
    profiles_test.go:162: Response:  {[{profile was created. success}]}
    profiles_test.go:162: Response:  {[{profile was created. success}]}
    profiles_test.go:162: Response:  {[{profile was created. success}]}
    profiles_test.go:162: Response:  {[{profile was created. success}]}
    profiles_test.go:162: Response:  {[{profile was created. success}]}
    profiles_test.go:162: Response:  {[{profile was created. success}]}
    profiles_test.go:162: Response:  {[{profile was created. success}]}
    profiles_test.go:162: Response:  {[{profile was created. success}]}
    profiles_test.go:162: Response:  {[{profile was created. success}]}
    profiles_test.go:162: Response:  {[{profile was created. success}]}
    profiles_test.go:162: Response:  {[{profile was created. success}]}
    statuses_test.go:96: Response:  {[{status was created. success}]}
    statuses_test.go:96: Response:  {[{status was created. success}]}
    statuses_test.go:96: Response:  {[{status was created. success}]}
    statuses_test.go:96: Response:  {[{status was created. success}]}
    statuses_test.go:96: Response:  {[{status was created. success}]}
    statuses_test.go:96: Response:  {[{status was created. success}]}
    statuses_test.go:96: Response:  {[{status was created. success}]}
    divisions_test.go:115: Response:  {[{division was created. success}]}
    divisions_test.go:115: Response:  {[{division was created. success}]}
    regions_test.go:89: Response:  {[{region was created. success}]}
    regions_test.go:89: Response:  {[{region was created. success}]}
    phys_locations_test.go:90: Response:  {[{physLocation was created. success}]}
    phys_locations_test.go:90: Response:  {[{physLocation was created. success}]}
    phys_locations_test.go:90: Response:  {[{physLocation was created. success}]}
    servers_test.go:245: Response:  atlanta-edge-01   {[{Server created success}]}
    servers_test.go:245: Response:  cdn2-test-edge   {[{Server created success}]}
    servers_test.go:245: Response:  influxdb02   {[{Server created success}]}
    servers_test.go:245: Response:  atlanta-router-01   {[{Server created success}]}
    servers_test.go:245: Response:  atlanta-edge-03   {[{Server created success}]}
    servers_test.go:245: Response:  atlanta-edge-14   {[{Server created success}]}
    servers_test.go:245: Response:  atlanta-edge-15   {[{Server created success}]}
    servers_test.go:245: Response:  atlanta-mid-16   {[{Server created success}]}
    servers_test.go:245: Response:  atlanta-org-1   {[{Server created success}]}
    servers_test.go:245: Response:  atlanta-org-2   {[{Server created success}]}
    servers_test.go:245: Response:  atlanta-mid-01   {[{Server created success}]}
    servers_test.go:245: Response:  rascal01   {[{Server created success}]}
    servers_test.go:245: Response:  edge1-cdn2   {[{Server created success}]}
    servers_test.go:245: Response:  riak   {[{Server created success}]}
    servers_test.go:245: Response:  denver-mso-org-01   {[{Server created success}]}
    servers_test.go:245: Response:  edge1-cdn1-cg3   {[{Server created success}]}
    servers_test.go:245: Response:  edge2-cdn1-cg3   {[{Server created success}]}
    servers_test.go:245: Response:  topology-edge-01   {[{Server created success}]}
    servers_test.go:245: Response:  topology-edge-02   {[{Server created success}]}
    servers_test.go:245: Response:  topology-mid-04   {[{Server created success}]}
    topologies_test.go:56: Response:  &{{a topology my-topology [{0 parentCachegroup [] <nil>} {0 secondaryCachegroup [] <nil>} {0 cachegroup3 [0 1] <nil>}] <nil>} {[{topology was created. success}]}}
    topologies_test.go:56: Response:  &{{another topology another-topology [{0 parentCachegroup [] <nil>} {0 cachegroup1 [0] <nil>} {0 secondaryCachegroup [] <nil>} {0 cachegroup2 [2] <nil>}] <nil>} {[{topology was created. success}]}}
    topologies_test.go:56: Response:  &{{A topology with secondary parents secondary-parents [{0 parentCachegroup [] <nil>} {0 cachegroup1 [0 2] <nil>} {0 secondaryCachegroup [] <nil>} {0 fallback1 [2 0] <nil>} {0 fallback2 [2 0] <nil>}] <nil>} {[{topology was created. success}]}}
    topologies_test.go:56: Response:  &{{A 4-tier topology 4-tiers [{0 parentCachegroup [] <nil>} {0 parentCachegroup2 [0] <nil>} {0 cachegroup1 [1] <nil>} {0 secondaryCachegroup [1 0] <nil>} {0 fallback1 [3] <nil>}] <nil>} {[{topology was created. success}]}}
    topologies_test.go:56: Response:  &{{This topology stems from 2 ancestors forked-topology [{0 topology-edge-cg-01 [2] <nil>} {0 topology-edge-cg-02 [6] <nil>} {0 topology-mid-cg-01 [3] <nil>} {0 topology-mid-cg-02 [4] <nil>} {0 topology-mid-cg-03 [5] <nil>} {0 topology-mid-cg-04 [] <nil>} {0 topology-mid-cg-05 [7] <nil>} {0 topology-mid-cg-06 [8] <nil>} {0 topology-mid-cg-07 [] <nil>}] <nil>} {[{topology was created. success}]}}
    deliveryservices_test.go:73: unable to get keys on cdn sslkeytransfer: 500 Internal Server Error[500] - Error requesting Traffic Ops https://localhost:6443/api/3.0/cdns/name/sslkeytransfer/sslkeys {"alerts":[{"text":"Internal Server Error","level":"error"}]}
    types_test.go:187: ---- DeleteTestTypes ----
--- FAIL: TestDeliveryServices (9.32s)
FAIL
FAIL	github.com/apache/trafficcontrol/traffic_ops/testing/api/v3	9.599s
FAIL
```
</details>

**Comment by zrhoffman** [comment]
Verified that TestDeliveryServices passes now!

**Review by mattjackson220 — APPROVED** [review]
looks great! docs build and look good, API tests pass, unit tests pass, manual testing works as expected!

**Pull request merged**